tests/frontend/test_tdmprogram.py

```diff
@@ -0,0 +1,425 @@
+    assert np.allclose(plusstdP2P3, squeezed_std, atol=atol)
+    minusstdP0X2 = (P0 - X2).std()
+    plusstdP0X2 = (P0 + X2).std()
+    expected = 2 * np.sinh(sq_r) ** 2
+    assert np.allclose(minusstdP0X2, expected, atol=atol)
+    assert np.allclose(plusstdP0X2, expected, atol=atol)
+    minusstdX3P1 = (X3 - P1).std()
+    plusstdX3P1 = (X3 + P1).std()
+    assert np.allclose(minusstdX3P1, expected, atol=atol)
+    assert np.allclose(plusstdX3P1, expected, atol=atol)
+
+
+def test_ghz():
+    """Generates a GHZ state and checks that the correct correlations (noise reductions) are observed
+    from the samples
+    See Eq. 5 of https://advances.sciencemag.org/content/5/5/eaaw4530
+    """
+    # Set up the circuit
+    np.random.seed(42)
+    n = 10
+    vac_modes = 1
+    copies = 1000
+    sq_r = 5
+    alpha = [np.arccos(np.sqrt(1 / (n - i + 1))) if i != n + 1 else 0 for i in range(n + vac_modes)]
+    alpha[0] = 0.0
+    phi = [0] * (n + vac_modes)
+    phi[0] = np.pi / 2
+
+    # Measuring X nullifier
+    theta = [0] * (n + vac_modes)
+    samples_X = singleloop(sq_r, alpha, phi, theta, copies)
+    reshaped_samples_X = np.array(samples_X).reshape([copies, n + vac_modes])
+
+    # We will check that the x of all the modes equal the x of the last one
+    nullifier_X = lambda sample: (sample - sample[-1])[vac_modes:-1]
+    val_nullifier_X = np.var([nullifier_X(x) for x in reshaped_samples_X], axis=0)
+    assert np.allclose(val_nullifier_X, 2 * np.exp(-2 * sq_r), rtol=5 / np.sqrt(copies))
+
+    # Measuring P nullifier
+    theta = [np.pi / 2] * (n + vac_modes)
+    samples_P = singleloop(sq_r, alpha, phi, theta, copies)
+
+    # We will check that the sum of all the p is equal to zero
+    reshaped_samples_P = np.array(samples_P).reshape([copies, n + vac_modes])
+    nullifier_P = lambda sample: np.sum(sample[vac_modes:])
+    val_nullifier_P = np.var([nullifier_P(p) for p in reshaped_samples_P], axis=0)
+    assert np.allclose(val_nullifier_P, n * np.exp(-2 * sq_r), rtol=5 / np.sqrt(copies))
+
+
+def test_one_dimensional_cluster():
+    """Test that the nullifier have the correct value in the experiment described in
+    See Eq. 10 of https://advances.sciencemag.org/content/5/5/eaaw4530
+    """
+    np.random.seed(42)
+    n = 20
+    copies = 1000
+    sq_r = 3
+    alpha_c = np.arccos(np.sqrt((np.sqrt(5) - 1) / 2))
+    alpha = [alpha_c] * n
+    alpha[0] = 0.0
+    phi = [np.pi / 2] * n
+    theta = [0, np.pi / 2] * (n // 2)  # Note that we measure x for mode i and the p for mode i+1.
+    reshaped_samples = np.array(singleloop(sq_r, alpha, phi, theta, copies)).reshape(copies, n)
+    nullifier = lambda x: np.array([-x[i - 2] + x[i - 1] - x[i] for i in range(2, len(x) - 2, 2)])[1:]
+    nullifier_samples = np.array([nullifier(y) for y in reshaped_samples])
+    delta = np.var(nullifier_samples, axis=0)
+    assert np.allclose(delta, 3 * np.exp(-2 * sq_r), rtol=5 / np.sqrt(copies))
+
+
+def test_two_by_n_cluster():
```

Comment:
The name of the function sounds like a 2D cluster state. Isn't the experiment about a 1D cluster state?

Reply:
changed it to `test_one_dimensional_cluster_tokyo`